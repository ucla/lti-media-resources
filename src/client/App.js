/* eslint-disable react/prop-types */
import React from 'react';
// before mounting your React application:
import theme from '@instructure/canvas-high-contrast-theme';
import './app.css';

import { Heading } from '@instructure/ui-heading';

theme.use({ overrides: { colors: { brand: 'red' } } });

const App = () => {
  const getLtikPromise = new Promise((resolve, reject) => {
    const searchParams = new URLSearchParams(window.location.search);
    let potentialLtik = searchParams.get('ltik');
    if (!potentialLtik) {
      potentialLtik = sessionStorage.getItem('ltik');
      if (!potentialLtik) reject(new Error('Missing lti key.'));
    }
    resolve(potentialLtik);
  });

  const setLtikPromise = new Promise((resolve, reject) => {
    getLtikPromise
      .then(res => {
        sessionStorage.setItem('ltik', res);
        resolve(res);
      })
      .catch(err => {
        reject(err);
      });
  });

  return <Heading>Hello World!</Heading>;
};

export default App;
