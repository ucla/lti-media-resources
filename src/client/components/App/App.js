/* eslint-disable react/prop-types */
import React from 'react';
// before mounting your React application:
import theme from '@instructure/canvas-high-contrast-theme';
import './app.css';

import { Heading } from '@instructure/ui-heading';

import { ltikPromise } from '../../services/ltik';

theme.use({ overrides: { colors: { brand: 'red' } } });

const App = () => <Heading>Hello World!</Heading>;

export default App;
