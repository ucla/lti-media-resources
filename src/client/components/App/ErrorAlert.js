import React from 'react';
import PropTypes from 'prop-types';

import { Alert } from '@instructure/ui-alerts';

export const ErrorAlert = ({ err, msg }) => {
  ErrorAlert.propTypes = {
    err: PropTypes.object,
    msg: PropTypes.string,
  };
  return (
    <Alert variant="error" renderCloseButtonLabel="Close">
      {`${msg}\n${err}`}
    </Alert>
  );
};
