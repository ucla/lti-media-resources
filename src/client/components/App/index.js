import React, { useState, useEffect } from 'react';
import { theme } from '@instructure/canvas-theme';
import { EmotionThemeProvider } from '@instructure/emotion'
import dompurify from 'dompurify';
import axios from 'axios';
import axiosRetry from 'axios-retry';

import { Tabs } from '@instructure/ui-tabs';
import { Bruincast } from '../Bruincast';
import { VideoReserve } from '../VideoReserve';
import { MusicReserve } from '../MusicReserve';
import { AdminPanel } from '../AdminPanel';
import { ErrorAlert } from './ErrorAlert';

import { getLtik } from '../../services/ltik';

const constants = require('../../../../constants');


axiosRetry(axios);

const App = () => {
  // Logic of changing tabs
  const [tabSelectedIndex, selectTabIndex] = useState(constants.TABS.BRUINCAST);
  const handleTabChange = (event, { index }) => {
    selectTabIndex(index);
  };

  // Declare states
  const [course, setCourse] = useState({});
  const [userid, setUserid] = useState(-1);
  const [onCampusStatus, setOnCampusStatus] = useState(true);

  const [bruincastCount, setBruincastCount] = useState(0);
  const [videoReserveCount, setVideoReserveCount] = useState(0);
  const [audioReserveCount, setAudioReserveCount] = useState(0);

  const [isAdmin, setIsAdmin] = useState(false);
  const [isInstructorOrAdmin, setIsInstructorOrAdmin] = useState(false);

  const [error, setError] = useState(null);

  // Get the current context (course and roles) from backend
  const retrieveContext = () => {
    const ltik = getLtik();
    axios
      .get(`${process.env.LTI_APPROUTE}/api/context?ltik=${ltik}`)
      .then((res) => {
        const { course: c, roles: r, userid: u, onCampus: oc } = res.data;
        setCourse(c);
        setUserid(u);
        setOnCampusStatus(oc);
        if (r && r.includes('administrator')) {
          setIsAdmin(true);
          setIsInstructorOrAdmin(true);
        }
        if (r && r.includes('instructor')) {
          setIsInstructorOrAdmin(true);
        }
        setError(null);
      })
      .catch((err) => {
        setError({
          err,
          msg: 'Something went wrong when retrieving app context...',
        });
      });
  };
  useEffect(retrieveContext, []);

  // Get the number of medias for each tab
  const retrieveNums = () => {
    const ltik = getLtik();
    axios
      .get(`${process.env.LTI_APPROUTE}/api/medias/counts?ltik=${ltik}`)
      .then((res) => {
        const { bruincasts, videos, audios } = res.data;
        setBruincastCount(bruincasts);
        setVideoReserveCount(videos);
        setAudioReserveCount(audios);
        setError(null);
      })
      .catch((err) => {
        setError({
          err,
          msg:
            'Something went wrong when getting the number of media entries...',
        });
      });
  };
  useEffect(retrieveNums, []);

  // Get notice from backend
  // Declaring function only; called in Bruincast component
  const [warning, setWarning] = useState('');
  // Bool that prevents retrieving warning from db every loads.
  const [retrievedWarning, setRetrievedWarning] = useState(false);
  const retrieveWarning = () => {
    // Already allowing hot notice update in front-end after submitting new notice in admin panel.
    // Hot notice update is faster than going to db so we only go to db once when there's no notice.
    if (!retrievedWarning) {
      const ltik = getLtik();
      axios
        .get(
          `${process.env.LTI_APPROUTE}/api/medias/bruincast/notice?ltik=${ltik}`
        )
        .then((res) => {
          setWarning(dompurify.sanitize(res.data));
          setRetrievedWarning(true);
          setError(null);
        })
        .catch((err) => {
          setError({
            err,
            msg: 'Something went wrong when retrieving bruincast notice',
          });
        });
    }
  };

  const videoReservesTabEnabled = () =>
    videoReserveCount > 0 || isInstructorOrAdmin;

  // Display admin tab only when 'roles' contains 'admin'
  let adminPanel = null;
  if (isAdmin) {
    adminPanel = (
      <Tabs.Panel
        id="adminPanel"
        renderTitle="Admin Panel"
        isSelected={tabSelectedIndex === constants.TABS.ADMIN_PANEL}
      >
        <AdminPanel
          warning={warning}
          setWarning={setWarning}
          retrieveNums={retrieveNums}
          setError={setError}
        />
      </Tabs.Panel>
    );
  }

  // JSX
  return (
    <EmotionThemeProvider theme={theme}>
      {error && <ErrorAlert err={error.err} msg={error.msg} />}
      <Tabs onRequestTabChange={handleTabChange}>
        <Tabs.Panel
          id="bruincast"
          renderTitle={`${
            constants.mediaTypeMap.get(constants.MEDIA_TYPE.BRUINCAST).string
          } (${bruincastCount})`}
          isSelected={tabSelectedIndex === constants.TABS.BRUINCAST}
        >
          <Bruincast
            course={course}
            warning={warning}
            retrieveWarning={retrieveWarning}
            userid={userid}
            isInstructorOrAdmin={isAdmin}
            setError={setError}
          />
        </Tabs.Panel>
        {videoReservesTabEnabled() && (
          <Tabs.Panel
            id="videoReserves"
            renderTitle={`${
              constants.mediaTypeMap.get(constants.MEDIA_TYPE.VIDEO_RESERVES)
                .string
            } (${videoReserveCount})`}
            isSelected={tabSelectedIndex === constants.TABS.VIDEO_RESERVES}
          >
            <VideoReserve
              course={course}
              onCampus={onCampusStatus}
              userid={userid}
              isInstructorOrAdmin={isInstructorOrAdmin}
              setError={setError}
            />
          </Tabs.Panel>
        )}
        <Tabs.Panel
          id="audioReserves"
          renderTitle={`${
            constants.mediaTypeMap.get(
              constants.MEDIA_TYPE.DIGITAL_AUDIO_RESERVES
            ).string
          } (${audioReserveCount})`}
          isSelected={
            tabSelectedIndex ===
            constants.TABS.DIGITAL_AUDIO_RESERVES -
              (!videoReservesTabEnabled() ? 1 : 0) // Reindex if VideoReserve tab is hidden
          }
        >
          <MusicReserve
            userid={userid}
            isInstructorOrAdmin={isInstructorOrAdmin}
            setError={setError}
          />
        </Tabs.Panel>
        {adminPanel}
      </Tabs>
    </EmotionThemeProvider>
  );
};

export default App;
