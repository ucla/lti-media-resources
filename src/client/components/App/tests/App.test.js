import React from 'react';
import axios from 'axios';
import { cleanup, render, waitFor } from '@testing-library/react';
import { shallow } from 'enzyme';
import App from '../index';

jest.mock('axios');
jest.mock('../../../services/ltik');

beforeEach(() => {
  document.documentElement.setAttribute('dir', 'ltr');
});
afterEach(cleanup);

test('Axios error handling in App', async (done) => {
  axios.get.mockImplementation(() =>
    Promise.reject(
      new Error({
        response: {
          status: 400,
        },
      })
    )
  );
  const { getByTestId } = render(<App />);
  const alertNode = await waitFor(() => getByTestId('alert'));
  expect(axios.get).toBeCalled();
  expect(alertNode).toBeDefined();
  done();
});

test('App component matches its original snapshot', (done) => {
  const appInstance = shallow(<App />);
  expect(appInstance).toMatchSnapshot();
  done();
});
