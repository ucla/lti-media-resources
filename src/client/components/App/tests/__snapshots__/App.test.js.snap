// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches its original snapshot 1`] = `
<Fragment>
  <Tabs
    elementRef={[Function]}
    onRequestTabChange={[Function]}
    shouldFocusOnRender={false}
    tabOverflow="stack"
    variant="default"
  >
    <Panel
      elementRef={[Function]}
      id="bruincast"
      isDisabled={false}
      isSelected={true}
      labelledBy={null}
      padding="small"
      renderTitle="Bruincast (0)"
      textAlign="start"
      variant="default"
    >
      <Bruincast
        course={Object {}}
        isInstructorOrAdmin={false}
        retrieveWarning={[Function]}
        setError={[Function]}
        userid={-1}
        warning=""
      />
    </Panel>
    <Panel
      elementRef={[Function]}
      id="audioReserves"
      isDisabled={false}
      isSelected={false}
      labelledBy={null}
      padding="small"
      renderTitle="Digital Audio Reserves (0)"
      textAlign="start"
      variant="default"
    >
      <MusicReserve
        isInstructorOrAdmin={false}
        setError={[Function]}
        userid={-1}
      />
    </Panel>
    <Panel
      elementRef={[Function]}
      id="mediaGallery"
      isDisabled={false}
      isSelected={false}
      labelledBy={null}
      padding="small"
      renderTitle="Media Gallery"
      textAlign="start"
      variant="default"
    >
      Media Gallery
    </Panel>
  </Tabs>
</Fragment>
`;
